File: library/src/main/java/com/qingmei2/library/base/fragment/BaseFragment.java
package com.qingmei2.library.base.fragment;

import android.app.Activity;
import android.app.ProgressDialog;
import android.databinding.DataBindingUtil;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.qingmei2.library.base.BaseViewModel;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;

/**
 * Created by QingMei on 2017/10/12.
 * desc:
 */

public abstract class BaseFragment<B extends ViewDataBinding, V extends BaseViewModel> extends Fragment {

    protected B binding;

    @Inject
    public V viewModel;

    protected View mRootView;

    protected ProgressDialog progressDialog;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        mRootView = LayoutInflater.from(getContext()).inflate(getLayoutRes(), container, false);
        return mRootView;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = DataBindingUtil.bind(view);
        initView(view);
        initData();
    }

    @Override
    public void onAttach(Activity activity) {
        AndroidSupportInjection.inject(this);
        super.onAttach(activity);
    }

    protected void onStateChanged(BaseViewModel.State state) {
        switch (state) {
            case LOAD_WAIT:
                break;
            case LOAD_ING:
                loading(true);
                break;
            case LOAD_SUCCESS:
                loading(false);
                break;
            case LOAD_FAILED:
                loading(false);
                break;
            default:
                break;
        }
    }

    protected void loading(boolean showing) {
        if (getContext() != null) {
            if (showing) {
                progressDialog = new ProgressDialog(getContext());
                progressDialog.setCancelable(false);
                progressDialog.show();
            } else if (progressDialog != null) {
                progressDialog.dismiss();
                progressDialog = null;
            }
        }
    }

    protected abstract int getLayoutRes();

    protected abstract void initView(View view);

    protected abstract void initData();

}

File: rhine/src/main/java/com/qingmei2/rhine/base/fragment/BaseFragment.kt
package com.qingmei2.rhine.base.fragment

import android.app.ProgressDialog
import android.databinding.DataBindingUtil
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.qingmei2.rhine.base.viewmodel.BaseRhineViewModel

abstract class BaseFragment<B : ViewDataBinding, V : BaseRhineViewModel> : Fragment() {

    protected var binding: B? = null

    var viewModel: V? = null

    protected lateinit var mRootView: View

    protected var progressDialog: ProgressDialog? = null

    protected abstract val layoutRes: Int

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mRootView = LayoutInflater.from(context).inflate(layoutRes, container, false)
        return mRootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding = DataBindingUtil.bind(view)
        initView(view)
        initData()
    }

    protected fun onStateChanged(state: BaseRhineViewModel.State) {
        when (state) {
            BaseRhineViewModel.State.LOAD_WAIT -> {
            }
            BaseRhineViewModel.State.LOAD_ING -> loading(true)
            BaseRhineViewModel.State.LOAD_SUCCESS -> loading(false)
            BaseRhineViewModel.State.LOAD_FAILED -> loading(false)
            else -> {
            }
        }
    }

    protected fun loading(showing: Boolean) {
        if (context != null) {
            if (showing) {
                progressDialog = ProgressDialog(context)
                progressDialog!!.setCancelable(false)
                progressDialog!!.show()
            } else if (progressDialog != null) {
                progressDialog!!.dismiss()
                progressDialog = null
            }
        }
    }

    protected abstract fun initView(view: View)

    protected abstract fun initData()

}
